### PR TITLE
Allow range expressions to be iterated through by for statements.

### LINKED ISSUE #951: Range as a for loop range
```ooc
range := 0 .. 5
for (i in range)
    println(i toString())
```
This doesn't build, because 
```
No such function iterator() for `Range`
```
What if rock could pre-process this into
```ooc
range := 0 .. 5
for (i in range min .. range max)
    println(i toString())
```
(which does work) so you wouldn't have to spell it out? As far as I can tell, a new `Range` is created with this syntax anyway.

### PR DESCRIPTION
**Pull request by shamanas**
Closes #951  
Also, sneakily fix the indexing in range for loops (the backend generated wrong code)

### CODE REVIEW

File: source/rock/middle/Foreach.ooc [hunk 1]
```diff
@@ -106,8 +101,75 @@ Foreach: class extends ControlStatement {
...
+
+                        // This is our new collection
+                        newCol := RangeLiteral new(VariableAccess new(access, "min", collection token),
+                                                   VariableAccess new(access, "max", collection token), collection token)
+
+                        collection = newCol
+                        replaced = false
+
+                        // We need to remove the unwrapped index vDecl from some parent scope
+                        // because it's type is unknown and currently unresolvable.
```
Comment (vendethiel) [inline source/rock/middle/Foreach.ooc]: its* type

Comment (shamanas) [inline source/rock/middle/Foreach.ooc]: oops

**Comment by vendethiel** [comment]
Cool stuff! :+1: (was there already a test for the fixed bug?)

**Comment by shamanas** [comment]
@vendethiel  
See the last two commits (test/compiler/control/foreach-range-generated.ooc)  
This is basically the most complex case (side-effects, index variable) :)

**Pull request merged**